python/spacetime/store.py

```diff
@@ -418,10 +432,37 @@ def register_app(self, app, typemap, name2class, name2baseclasses):
             final_base_types = [b for b in base_types if b.__PCC_BASE_TYPE__]
             self.__base_store.add_types(final_base_types)
 
-    def gc(self, app, name2class):
+    def gc(self, this_app, name2class):
+        mylock = self.__copylock[this_app]
+        self.pause()
         self.logger.warn("Application %s disconnected. Removing owned objects.",
-                         app)
-        for strtp in self.__gc[app]:
-            for oid in self.__gc[app][strtp]:
+                     this_app)
+
+        for strtp in self.__type_to_app.keys():
+            if this_app in self.__type_to_app[strtp]:
+                self.__type_to_app[strtp].remove(this_app)
+
+        other_apps = set()
+        for strtp in self.__gc[this_app]:
+            if strtp in self.__base_store.get_base_types():
+                if strtp in self.__type_to_app.keys():
+                    other_apps = set(self.__type_to_app[strtp]).difference(set([this_app]))
+
+        # delete all owned objects, and inform other simulations of deleted objects
+        for strtp in self.__gc[this_app]:
+            for oid in self.__gc[this_app][strtp]:
                 self.__base_store.delete(name2class[strtp], oid)
-        del self.__gc[app]
+                for app in other_apps:
```

Comment:
oh man. Renaming app again. Damn. The feel the need for one time assignment like in golang now.